dissect/hypervisor/disk/hdd.py

```diff
@@ -0,0 +1,392 @@
+from __future__ import annotations
+
+from bisect import bisect_right
+from dataclasses import dataclass
+from functools import cached_property
+from pathlib import Path
+from typing import BinaryIO, Iterator, Optional, Tuple, Union
+from uuid import UUID
+from xml.etree.ElementTree import Element
+
+try:
+    from defusedxml import ElementTree
+except ImportError:
+    from xml.etree import ElementTree
+
+from dissect.util.stream import AlignedStream
+
+from dissect.hypervisor.disk.c_hdd import SECTOR_SIZE, c_hdd
+from dissect.hypervisor.exceptions import InvalidHeaderError
+
+DEFAULT_TOP_GUID = UUID("{5fbaabe3-6958-40ff-92a7-860e329aab41}")
+
+
+class HDD:
+    """Parallels HDD virtual disk implementation.
+
+    Args:
+        path: The path to the .hdd directory or .hdd file in a .hdd directory.
+    """
+
+    def __init__(self, path: Path):
+        if path.is_file() and path.parent.suffix.lower() == ".hdd":
+            path = path.parent
+        self.path = path
+
+        descriptor_path = path.joinpath("DiskDescriptor.xml")
+        if not descriptor_path.exists():
+            raise ValueError("Invalid Parallels HDD path (missing DiskDescriptor.xml)")
+
+        self.descriptor = Descriptor(descriptor_path)
+
+    def _open_image(self, path: Path) -> BinaryIO:
+        """Helper method for opening image files relative to this HDD.
+
+        Args:
+            path: The path to the image file to open.
+        """
+        root = self.path
+        filename = path.name
+
+        if path.is_absolute():
+            if not path.exists():
+                # File is in same directory
+                candidate_path = root / filename
```

Comment:
Added more comments, is it clearer now?

Reply:
Yes, but I think if you'd add how `root` and `path` start out before they are merged it would be even more clear.

So for the first case that would be (I think):
`root`: `/some/path/example.pvm/example.hdd/`
`path`: `/example.pvm/example.hdd/example.ext`
`candidate_path`: `/some/path/example.pvm/example.hdd/example.ext`

Reply:
Added those as well as comments.